src/fields/ntt.rs

```diff
@@ -0,0 +1,268 @@
+    /// the 2^i-th root of unity. It should start with 1 itself, and contain at least enough values
+    /// to include the `values.len()`-th root of unity. Note that for each element in the array, its
+    /// predecessor is its square.
+    ///
+    /// # Panics
+    ///
+    /// This panics if the length of the input is not a power of two.
+    ///
+    /// This panics if there are not enough roots of unity in `omegas`.
+    fn ntt_bit_reversed(values: &mut [Self], omegas: &[Self]) {
+        let log_n = usize::try_from(values.len().ilog2()).unwrap();
+        if 1 << log_n != values.len() {
+            panic!(
+                "length of input to NTT was {}, which is not a power of two",
+                values.len()
+            );
+        }
+
+        // Evaluate the NTT with the decimation-in-frequency radix-2 FFT algorithm.
+        let mut stride = 1 << (log_n - 1);
+        for omega in omegas[1..=log_n].iter().rev() {
+            // The i=0 iteration of the below loop is unrolled separately to save some multiplications.
+            let mut j = 0;
+            while j < values.len() {
+                (values[j], values[j + stride]) = (
+                    values[j] + values[j + stride],
+                    (values[j] - values[j + stride]),
+                );
+
+                j += stride * 2;
+            }
+
+            let mut omega_power = *omega;
+            for i in 1..stride {
+                let mut j = i;
+                while j < values.len() {
+                    (values[j], values[j + stride]) = (
+                        values[j] + values[j + stride],
+                        (values[j] - values[j + stride]) * omega_power,
+                    );
+
+                    j += stride * 2;
+                }
+                if i < stride - 1 {
+                    omega_power *= *omega;
+                }
+            }
+
+            stride /= 2;
+        }
+    }
+
+    /// Computes the inverse Number Theoretic Transform of a sequence. The input must be in
+    /// bit-reversed order. The result is returned in-place in the natural order.
+    ///
+    /// The `omega_inverses` argument must be a list of power-of-two roots of unity, such that
+    /// element i is the 2^i-th root of unity. It should start with 1 itself, and contain at least
+    /// enough values to include the `values.len()`-th root of unity. Note that for each element in
+    /// the array, its predecessor is its square. These values should be the multiplicative inverses
+    /// of the roots of unity used during the forwards NTT transformation.
+    ///
+    /// The `size_inv` argument must be the multiplicative inverse of the length of the input array.
+    ///
+    /// # Panics
+    ///
+    /// This panics if the length of the input is not a power of two.
+    ///
+    /// This panics if there are not enough roots of unity in `omegas`.
+    fn inverse_ntt_bit_reversed(values: &mut [Self], omega_inverses: &[Self], size_inv: Self) {
```

Comment:
Why not have this function compute `size_inv` from `values.len()`? Do we expect that the caller will be able to precompute it faster than this function could inject the length into the field and call `mul_inv`?